Add named arg parser

jpbetz opened the pull request:
Lots going on here:

- Add new ExtractFunctionStyleCommentTypedTags that supports both typed args and named args.(thanks @thockin for the parser!)
- All Extract functions call down to ExtractFunctionStyleCommentTypedTags to keep code paths simple.  But only the new function supports named params.
- Expand test coverage.
- Remove bare JSON support.

jpbetz commented:
/assign @thockin 
This makes a few small additions to the parser.  Namely support for tag keys that end in `=` followed by a value and support for the `:` char in tag names.

jpbetz commented on v2/argsparser.go:
```diff
@@ -99,9 +108,13 @@ parseLoop:
 			case isIdentInterior(r):
 				tag.WriteRune(r)
+			case r == ':': // allowed in tag names
+				tag.WriteRune(r)
```
@thockin We could expand the grammar from `<tagBody> ::= <name> ...` to `<tagBody> ::= { <prefix> ":" } <name> ...`.  But what I did here was to just allow ":" in tag names.

Thoughts welcome.

thockin replied:
Do we have a use-case for it?

jpbetz replied:
We have to somehow support `+k8s:<name>`. My test case for such tags was failing with the grammar.

jpbetz commented on v2/argsparser.go:
```diff
@@ -117,6 +130,9 @@ parseLoop:
 				st = stPrefixedNumber
+			case r == '-':
+				buf.WriteRune(r)
+				st = stNumber
```
@thockin Should we allow negative numbers?  I've added it here.

thockin replied:
Do we have a use-case?  If we negative, do we also need '+' ?

Also I would write this as:

```
case r == '-' || r == '+' || unicode.IsDigit(r):
    buf.WriteRune(r)
    st = stNumber
```

...just to save on space

jpbetz replied:
Don't need negatives today, but since we are handling hex and octal, I threw it in for completeness.

Added + support.

jpbetz commented on v2/argsparser.go:
```diff
@@ -222,9 +238,15 @@ parseLoop:
+		case stTrailingValue:
+			valueStart = i
+			i = len(runes) - 1
+			break parseLoop
```
@thockin The idea here was to track where the "beginning of the key value" is found so I can return the position of it in the parse result.

thockin replied:
Alternate approach, accumulate all bytes, and when you hit the end of the loop `if st == stValue` then save that and return it.  Downside: any trailing comment is still there.  If someone calls this func again, it will "just work"

jpbetz replied:
Agreed.  I'll leave the trailing comment in the value and parse it as opaque text.  See below comments for more detail.

jpbetz commented on v2/comments.go:
```diff
@@ -188,6 +185,19 @@ func ExtractFunctionStyleCommentTags(marker string, tagNames []string, lines []s
+	Args []string
+	// Value is the value of the tag.
+	Value string
+}
```
This is the old tag representation.  Kept for backward compat

jpbetz commented on v2/comments.go:
```diff
@@ -249,12 +246,17 @@ type TypedTag struct {
-	Typ   ArgType
+	// Type is the type of the Value.
+	Type ArgType
 }
```
Is this string + type representation sufficient?  We could take this further and have types like:

```go
type String string

type (String) Type() ArgType {
  return TypeString
}
```

@aaron-prindle for thoughts on processing this downstream.

thockin commented on v2/comments.go:
```diff
@@ -213,6 +239,30 @@ type Tag struct {
+	Typ   ArgType
+}
+
+type ArgType string
```
godoc

thockin commented on v2/comments.go:
```diff
@@ -105,12 +105,11 @@ func ExtractSingleBoolCommentTag(marker string, key string, defaultVal bool, lin
-// JSON syntax. If not specified (either as "key=value" or as
-// "key()=value"), the resulting Tag will have an empty Args list.
+// The arg is optional. It may be a Go identifier, a raw string literal enclosed
+// in back-ticks, an integer (with support for hex, octal or binary notation) or
```
back-ticks or doubel-quotes

thockin commented on v2/comments.go:
```diff
@@ -251,32 +268,49 @@ type Arg struct {
-// ArgType is the type of an arg.
-type ArgType string
+func (i Int) String() string {
+	return strconv.Itoa(int(i))
```
Why are you going thru the effort to parse string->int and then return it as a string anyway?

I'm not sure I see the value of the extra abstraction here

jpbetz replied:
It made more sense when all of the types were backed by a string (`type Bool string`) :(

Removing.

jpbetz commented on v2/comments.go:
```diff
@@ -178,26 +197,33 @@ func ExtractFunctionStyleCommentTags(marker string, tagNames []string, lines []s
 
-		name, args, err := s.parseTagKey(tagNames)
+		line = stripTrailingComment(line)
+		kv := strings.SplitN(line, "=", 2)
```
Since a value can be another tag, or opaque text.  I think the only fix for trailing comments is to make the caller deal with it.. I'll drop the trailing comment removal.

I'll have the parser consume everything else and return it.

jpbetz commented:
Thanks @thockin.  Excellent feedback.  I've applied it. I think this is getting close now.

I'm not happy with having to drop trailing comment support, but I think it's better to drop it for now and only add support if we can come up with a clean solution.

jpbetz commented:
@thockin All feedback applied.  This is now sufficient for extracting and parsing comments and also parsing nested tags that don't have a marker/group prefix ("+k8s:").

thockin commented on v2/comments.go:
```diff
@@ -229,227 +225,185 @@ func (t Tag) String() string {
+		if idx := strings.IndexAny(line, "(="); idx > 0 {
+			nameEnd = idx
+		}
+		ident := TagIdentifierFromString(line[:nameEnd])
```
If we keep "group", we could just move this to the state-machine -- would it be easier?

jpbetz replied:
I've dropped the group concept and allowed ":" in tag names.

thockin commented on v2/comments.go:
```diff
@@ -229,227 +225,185 @@ func (t Tag) String() string {
-		if err != nil {
-			return 0, 0, err
+// ExtractAndParseTagWithArgs combines ExtractTags and ParseTagWithArgs.
+func ExtractAndParseTagWithArgs(marker string, group *string, lines []string) (map[TagIdentifier][]TypedTag, error) {
```
cute composition, but how much of it needs to be public API?

jpbetz replied:
Dropped it.  I kept a `ParseAll` utility. Let me know if even that is too much.

thockin reviewed (commented):
I wonder if we should make a sub-pkg for this now?

We tend to use "tags" which could be a good pkg name, or "directive" ?

Then you can pick a good name for types and functions, giving the old versions less nice names, and leaving a forwarding function from the root pkg.

gengo/v2/tags.ExtractCommentTags is `var ExtractCommentTags = tags.ExtractLegacySimpleTags` :)

jpbetz commented on v2/argsparser.go:
```diff
@@ -0,0 +1,272 @@
+			case r == '(':
+				incomplete = true
+				st = stArg
+			case r == '=':
```
I've added back trailing comment support. It's only allowed if there is no value part (for soundness of the parser).

thockin commented on v2/codetags/types.go:
```diff
@@ -0,0 +1,147 @@
+// MustIntValue parses an int argument. Panics if the argument is invalid.
+func MustIntValue(s string) IntValue {
+	i, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		panic(fmt.Sprintf("invalid int value: %q", s))
+	}
+	return IntValue{s: s, i: i}
+}
+
+func (a IntValue) String() string {
+	return a.s
+}
+
+// Int returns the integer value of the argument.
+func (a IntValue) Int() int64 {
+	return a.i
+}
+
+func (a IntValue) Type() Type {
+	return TypeInt
+}
+
+type BoolValue bool
+
+func (a BoolValue) String() string {
+	if a == true {
+		return "true"
+	} else {
+		return "false"
+	}
+}
+
+// Bool returns the boolean value of the argument.
+func (a BoolValue) Bool() bool {
+	return bool(a)
+}
+
+func (a BoolValue) Type() Type {
+	return TypeBool
+}
```
I'll be honest, I still don't see the point of the abstraction here, but I am not going to argue it further.

This just feels like a fancy way of saying:

```
type ArgType string

const (
    TypeInt ArgType = "int"
    TypeBool ArgType = "bool"
    TypeString ArgType = "string"
)

type Arg stuct {
    Name string
    Type ArgType
    Value string
}
```

...without offering a whole lot more safety.

jpbetz replied:
Apologies.  I'be been searching for ways to minimize the amount of parsing callers need to do.  This feedback suggests I've over-fit for that. I'll back this out.

thockin replied:
I feel like I have come full circle in my Go journey - I initially looked for places to build abstract and I have now reached a point where I am trying to take it out. :)

thockin commented on v2/codetags/types.go:
```diff
@@ -0,0 +1,147 @@
+// TypedValue represents a parsed value.
+// It may be a string, int, or bool. TypedValue retains the original string representation
+// of the value that was parsed from the tag.
+type TypedValue interface {
```
TypedValue makes it sound like the value of the tag, not the value of the arg.  See my other comment about abstraction.  Naming things is hard and you have made two things to name instead of just one.

thockin commented on v2/codetags/parser.go:
```diff
@@ -0,0 +1,372 @@
+			i = len(runes) - 1
+			break parseLoop
+		default:
+			panic("unknown state")
```
I know it's kind an `ASSERT()` but it still feels dirty to panic.

And anyway, you could add the string of the unhandled state :)

thockin commented on v2/comments.go:
```diff
@@ -74,6 +73,10 @@ func ExtractCommentTags(marker string, lines []string) map[string][]string {
+//
+// This function is a wrapper around codetags.Extract and codetags.Parse, but only supports tags with
+// a single position arg of type string, and a value of type bool.
+// This function uses the full group:name identifier as the returned Tag.name.
```
references "group"

thockin commented on v2/comments.go:
```diff
@@ -92,117 +95,57 @@ func ExtractSingleBoolCommentTag(marker string, key string, defaultVal bool, lin
-// This function should be preferred instead of ExtractCommentTags.
+// This function is a wrapper around codetags.Extract and codetags.Parse, but only supports tags with
+// a single position arg of type string.
+// This function uses the full group:name identifier as the returned Tag.name.
```
references "group"

thockin commented on v2/codetags/extractor.go:
```diff
@@ -0,0 +1,60 @@
+//
+//	map[string][]string{
+//		"required": {"required"},
+//		"required": {"format=k8s-long-name"},
```
why is the key here "required" ?

jpbetz replied:
copy-pasta.  Fixing.

thockin commented on v2/codetags/extractor.go:
```diff
@@ -0,0 +1,60 @@
+// Example: When called with prefix "+k8s:", lines:
+//
+//	Comment line without marker
+//	+k8s:required
```
Can you include a (fake) example which explains why it returns `[]string` ?

thockin replied:
Also mayber an example with args to show why the map-key is included in the value?  E.g. I think it would be somewthing like:

```
"withArgs": {"withArgs(abc123)=value"},
```

right?

jpbetz replied:
SG.  I found a bug in Extract when expanding the test coverage. Fixing that too.  The example now covers this functions capabilities more extensively.

thockin commented on v2/comments.go:
```diff
@@ -46,7 +45,7 @@ import (
 //	map[string][]string{"foo":{"value1, "value2"}, "bar": {""}, "baz": {`"qux"`}}
 //
-// Deprecated: Use ExtractFunctionStyleCommentTags.
+// Deprecated: Prefer codetags.Extract, codetags.Parse and codetags.ExtractAndParse.
```
ExtractAndParse no longer exists?

thockin reviewed (commented):
mostly small things

jpbetz commented:
@thockin Feedback applied.  I've reverted back to the type field approach.  I found a bug in extract so I've cleaned that code up. Most of the rest of the work is test and godoc cleanup.

jpbetz commented on v2/codetags/extractor.go:
```diff
@@ -49,12 +60,29 @@ func Extract(prefix string, lines []string) map[string][]string {
-			nameEnd = idx
+		nameEnd := findNameEnd(line)
+		if nameEnd == 0 {
+			continue
```
Keeping it SGTM. Updated.  I've also refined the whitespace handling to preserve the value as much as possible. This feels inline with our goal of simplifying these functions down to a single purpose.

jpbetz commented:
@thockin feedback applied

yongruilin commented on v2/codetags/parser_test.go:
```diff
@@ -17,22 +17,32 @@ limitations under the License.
-func TestParseTagKeyWithTagNames(t *testing.T) {
-	mkss := func(s ...string) []string { return s }
+func TestParse(t *testing.T) {
+	mkss := func(s ...string) []Arg {
```
nit: mkss -> mksa ?

jpbetz commented on v2/codetags/parser_test.go:
```diff
@@ -45,31 +55,72 @@ func TestParseTagKeyWithTagNames(t *testing.T) {
+		// invalid args
+		{input: "name(arg1, arg2)", err: true},
+		{input: "argWithDash(arg-name) ", err: true},
+		{input: "withRaw(`a = b`)", expectKey: "withRaw", expectArgs: mkss("a = b")},
```
Updated

jpbetz commented:
/hold
For @thockin LGTM for merge.

thockin commented:
FWIW, we should probably fix any places that accept payloads and clarify that the args are strings.  No more naked JSON.

jpbetz commented:
> FWIW, we should probably fix any places that accept payloads and clarify that the args are strings.  No more naked JSON.

We haven't accumulated any uses in k/k yet (I just double checked).  The only use is in the validation-gen dev branch and we can easily fix.  We'll follow up in the clarification in godoc too.

jpbetz commented on v2/codetags/parser.go:
```diff
@@ -47,56 +47,78 @@ type tagKey struct {
+// For example,
+//
+//	"key // This tag has no value, so this comment is ignored"
+//	"key=value // Comments after values are treated as part of the value"
```
I didn't think about this long enough before responding.  Opening a convo on chat to hammer it down.

thockin commented:
I see a bunch of tests that use naked JSON with `validateFalse` ?
